WIP: Extract useful information from connection url

shulard opened the pull request:
Work on the #25 issue about URL details extraction.

This PR must allow the library to extract connection URL details and use it during connection :

* Entity name to join
* Credentials
* Connection options
* Connection flags (used to define entity more precisely)

Example :

```
$uri    = 'irc://user:pass@irc.efnet.org:194/#channel?key=abcd';
$client = new Hoa\Irc\Client(new Hoa\Socket\Client($uri));
```

The given code is valid and can be used to join the channel `#channel` with the option `key=abcd`.

Hywan commented:
Hello @shulard!

Is it ready for a review?

shulard commented:
Hello @Hywan,

I haven't worked on `Connection` implementation for the moment, only on URL information extraction. You can take a look if you want, this first part is ready for a review 😄.

Hywan commented on Socket.php:
```diff
@@ -49,21 +49,201 @@
+        $this->_username = $username;
+        $this->_password = $password;
+        $this->_entity   = $entity;
+        //TODO: Maybe authorize only valid options like described in:
```
Will you address this todo in this PR? If no, consider to open another issue.

Hywan replied:
Will you address this todo in this PR? If no, consider to open another issue.

shulard replied:
I'll address it in this PR, not sure what to do for the moment here...

Hywan commented on Socket.php:
```diff
@@ -49,21 +49,201 @@
+     *
+     * @const string
+     */
+    const ENTTYPE_ISUSER = 'isuser';
```
Please, do not use shortcuts or acronyms, e.g. `USER_ENTITY`, `CHANNEL_ENTITY` would be fine I guess.
Also, I am not sure they are “entities”, but more things like “parts”. Is entity the wording of the RFC? If yes, keep it.

Hywan commented on Socket.php:
```diff
@@ -49,21 +49,201 @@
+     *
+     * @const string
+     */
+    const ENTTYPE_ISCHANNEL = 'ischannel';
```
I think you can use integers instead strings here. It will be lighter. And since we have constants, it does not matter.

Hywan replied:
I suggest `const ENTITY_TYPE_CHANNEL` = 42;` :-)

Hywan commented on Socket.php:
```diff
@@ -49,21 +49,201 @@
+     * Entity type.
+     * @var string
+     */
+    private $_entityType = self::ENTTYPE_ISCHANNEL;
```
Be sure this works in PHP 5.5.

Hywan commented on Socket.php:
```diff
@@ -49,21 +49,201 @@
+     * Host type.
+     * @var string
+     */
+    private $_hostType;
```
Any reason to put everything in the private scope? By default, we prefer to use the protected scope, except strong reasons.

shulard replied:
No real reason, just a habit 😄, I'll update scope.

Hywan commented on Socket.php:
```diff
@@ -49,21 +49,201 @@
+    private $_hostType;
+
+    /**
+     * Options list.
```
Please, give more details here.

Hywan commented on Socket.php:
```diff
@@ -49,21 +49,201 @@
+     * @param   string   $entity      Entity to access directly.
+     * @param   string   $username    Username to log in.
+     * @param   string   $password    Password to authenticate user.
+     * @param   array    $flags       List of flags to caracterize the entity.
```
```diff
- caracterize
+ characterize
```

Hywan replied:
Also, provide some examples (or refer to constants if any).

Hywan commented on Socket.php:
```diff
@@ -49,21 +49,201 @@
+        $entity        = null,
+        $username      = null,
+        $password      = null,
+        array $flags   = null,
```
```diff
- array $flags = null
+ array $flags = []
```
?

Hywan commented on Socket.php:
```diff
@@ -49,21 +49,201 @@
+        $this->_options  = $options;
+
+        if (null === $this->_entity && null !== $flags) {
+            throw new Exception("Can't define flags without defining entity.");
```
We prefer to use “Cannot” instead of “Can't”. It avoids a quote :wink:.

Hywan commented on Socket.php:
```diff
@@ -49,21 +49,201 @@
+        }
+        if (count($flags) > 2) {
+            throw new Exception(
+                "Can't have more than two flags [enttype, hosttype]."
```
Why flags in plural then?

Hywan commented on Socket.php:
```diff
@@ -49,21 +49,201 @@
+                "Can't have more than two flags [enttype, hosttype]."
+            );
+        }
+        while (count($flags) > 0) {
```
We can have 0 or 1 flag, so why a loop?

Hywan commented on Socket.php:
```diff
@@ -100,9 +280,44 @@ public static function transportFactory($socketUri)
+        $flags    = null;
+        $options  = [];
+
+        $pattern  = '/^((#|%23)?([^,?]+))(,([^?]+))?(\?(.*))?$/';
```
Why `%23`? It looks like it is encoded.

Hywan replied:
Maybe you should explicitly escape ? (in `[^,\?]` for instance).

Hywan replied:
You can name pattern in the regular expression, thus you will be able to refer to them with explicit name instead of abstract index. For instance: `/^(?<foo>\w+/` and then `$matches['foo']` instead of `$matches[1]`.

shulard replied:
In the RFC, it's mentioned that a client must match an encoded # if present... Maybe I understand it wrong : https://tools.ietf.org/html/draft-butcher-irc-url-04#section-2.5.1

shulard replied:
Ok I'll use it, it allows to be more understandable 😄.

Hywan replied:
This is correct, I understand this too.

shulard commented:
Hello @Hywan,

Sorry for the delay but I've worked on that PR this week 😄.

Now, if you use something like that, it works (User is used and Channel is joined) :

```php
$socket = new Hoa\Socket\Client('irc://Gordon:password@chat.freenode.net/#hoaproject-random');
$client = new Hoa\Irc\Client($socket);
$client->run();
```

It works with secured account, guest account. I still need to handle some options / flags, I make that changes ASAP.

I've used event handlers (on `open`) to set specific connection details, do you think it's the right thing to do ?

Hywan commented:
Please, stop being sorry for “delay”. This is open source. We all have a life, with different constraints. I am the first to be concerned…

THANK YOU for your contribution!

Hywan commented:
Yes, the `open` listener is the best one to call the `join` method I guess.

shulard commented:
Hello,

I've pushed new changes to handle all the described properties.

Now you can use that kind of code :

```php
//Ask to connect as "gordon" then join channel
$socket = new Hoa\Socket\Client('irc://chat.freenode.net/gordon,isuser');
$client = new Hoa\Irc\Client($socket);
$client->on('open', function($bucket) {
    $bucket->getSource()->setChannel('#hoaproject-random');
});
$client->run();

//Directly connect on the "#hoaproject-random" channel as "gordon"
$socket = new Hoa\Socket\Client('irc://gordon:@chat.freenode.net/#hoaproject-random');
$client = new Hoa\Irc\Client($socket);
$client->run();

//Connect using a protected username on "#hoaproject-random" channel
$socket = new Hoa\Socket\Client('irc://gordon:password@chat.freenode.net/#hoaproject-random');
$client = new Hoa\Irc\Client($socket);
$client->run();
```

I've also updated unit tests, removed TODOs in the code (because they are done now) and used named groups in regex.
Maybe the commit history needs a little cleanup but the code is working for me.

In `Socket.php` I've added a new regex to parse a nickname with optional part. I saw that there is one in `Client.php` but the 3 parts are mandatory (nick, user, host). Maybe we can merge the two in a helper function, I don't know if they must be mandatory or not 😄.
